### ISSUE TITLE
Testing process is non-streamlined and unintuitive

### ISSUE THREAD

**Issue by OrdoFlammae** [triplea-game/triplea#5019]
This is a minor issue, but I think there should be a better way to test the project before submitting pull requests. Right now, the process involves starting a docker process and a test http-server before running a shell-script that runs a bunch of checks. I see a few problems with this process. First, I can't do it inside my IDE, which means that every time I want to check something, I have to navigate outside of the IDE to do it, which breaks up the development process. Secondly, following from not having it in the IDE, I have to have multiple windows running around just to test the code. If we had the build process inside the IDE, it would be easier to run the multiple processes. Lastly, why are we using shell scripts to do this when we can just make another Gradle configuration to do it? That is what Gradle is for, right?

Ultimately, what I would like would be one Gradle task that runs through all the other necessary Gradle tasks, and does the whole process. Or maybe two, if one is too impractical, one to start up the server, and one to run all the tests (in which case, only one really needs to be implemented, the testing task will be simple).

I will be working on this in my spare time, but I would like some input from the development team who have been working on this project for a lot longer than me (ha ha). Do you think this is a good idea? Would you have any suggestions for implementing this?

**Comment by RoiEXLab** [comment]
For unit tests alone you shouldn't have to leave your IDE at all. You can create a test config in IntelliJ IDEA and use that to run all the unit tests, not that you'll have to exclude JUnit 5 tags called "integration" for this to exclude integration tests that require a postgresql service running in the background with a database called ta_users IIRC, everything else should work there no problem.
 I must admit that I mostly rely on the travis build to run all the tests, but during development I find it more convinient to just run invididual classes directly from the IDE anyways.

You have a point though

**Comment by DanVanAtta** [comment]
Testing of TripleA I would suggest is a critical issue. 


There are some interesting papers including TripleA (though somewhat dated and they look at TripleA from when it was hosted in sourceforge). One paper essentially finds that our tests have a number of problems (91% of tests contain test smells and it finds there is a correlation between test smells and maintenance effort):
- http://www.cs.loyola.edu/~binkley/papers/icsm-12-smells.pdf

Another defines metrics to find poor testing means the codebase is inefficient to work with (unmaintainable and brittle): https://testroots.org/assets/papers/athanasiouTSE.pdf


> (Page 17) 6.3.2  Hypothesis 2 : The relation between test code qualityand throughput

> If   we   consider   defect   throughput   and   enhancementthroughput separately, we see that both are significantly cor-related with test code quality at the 99% confidence level.The correlation is stronger for enhancement throughput

> (Page 18) throughput  and  productivity  confirm our  expectation  that  they  are  related  to  the  quality  of  testcode.


In practice, TripleA faces a near existential problem here. The level of effort for maintenance has been and arguably is greater than the effort than can be provided by the maintainers. That means we spend all our time on community communication, manual testing, bug fixes, operations overhead, and so much so that we can't move the project forward. I believe that was the case 5 years ago. For example, it was a full time job to moderate the TripleA forums, a full time job to manage the maps, a full time job to manage the code base, a near full time job to manage the servers, and a multi-hour effort to test the smallest of changes.

With luck we've greatly improved that equation over the last 3~5 years and can make better forward progress today and maybe can start to really enhance the game.


***TL;DR***: testing is critical to TripleA and improving it is vital to project success

--------

@OrdoFlammae 

>  First, I can't do it inside my IDE, which means that every time I want to check something, I have to navigate outside of the IDE to do it

Once docker and the http-server are running, you should be able to run tests from within IDE any number of times. So in essence launching those background operations lets you then run tests any number of times. It is not ideal.


> Secondly, following from not having it in the IDE

IDEA does support launching docker containers, that could be a possibility. Having a gradle command to start docker would be good.

> Lastly, why are we using shell scripts to do this when we can just make another Gradle configuration to do it?

Gradle can be used to launch the http-server, the shell script AFAIK is just a call to gradle. The shell scripts are in part there to document everything that needs to be done and to make it more or less one command.

The http-server gradle run command is blocking. I don't know how to get it so that gradle can launch the http-server in background and continue with other tasks.

The answer you're perhaps really looking for is that we simply aren't there yet with more testing automation/efficiency. We've come some ways to make testing easier, clearly it can be better and more effort/progress is highly desirable.


> Or maybe two, if one is too impractical, one to start up the server, and one to run all the tests (in which case, only one really needs to be implemented, the testing task will be simple).

If we can launch docker containers from gradle, have it not be blocking, and survive for a subsequent gradle run, then getting the DB up and running could be integrated. By the same token, launching another docker to then run the http-server should also be feasible. At that point we could have it all be a set of chained gradle operations and it would then be one gradle command : )


> I will be working on this in my spare time, but I would like some input from the development team who have been working on this project for a lot longer than me (ha ha). Do you think this is a good idea? Would you have any suggestions for implementing this?

Great to hear! The effort is super welcome. Feel free to bounce ideas off of us.

**Comment by RoiEXLab** [comment]
> Once docker and the http-server are running, you should be able to run tests from within IDE any number of times. So in essence launching those background operations lets you then run tests any number of times. It is not ideal.

Side note: In intellij you can add "build steps" to your run configuration and hide console output of those build steps. If this build step runs `docker start <containerid>` then running a test will silently launch a container in the background without you eceneven noticing.
Of course this doesn't solve all problems (I mean clicking on tests creates a completely new run configuration without those settings, so doesn't solve that case) but it greatly helps preventing the "my bad, forgot to launch the database"-moment when launching something.
I have been using it quite some time now for a spring boot project at work and it works pretty good.

**Comment by OrdoFlammae** [comment]
I am looking at docker plugins for gradle. Right now, I am leaning towards [bmuschko](https://github.com/bmuschko/gradle-docker-plugin)'s (well documented, used more than others, etc.) Do y'all have any ideas?

**Comment by RoiEXLab** [comment]
I never used gradle with docker, so can't offer any advice here.

Go ahead if you think that's a good plugin choice

**Comment by DanVanAtta** [comment]
@OrdoFlammae please re-open if there is anything more to discuss; closing for now as there seem to be no further action items.
